src/ZVal.h

```diff
@@ -0,0 +1,505 @@
+	 * @return  True on success, false if 'v' has never been set to
+	 * a value (which we can only tell for managed types).
+	 */
+	bool CopyElement(unsigned int n, const ZAMValUnion& v)
+		{
+		if ( zvec.size() <= n )
+			GrowVector(n + 1);
+
+		if ( managed_yt )
+			return SetManagedElement(n, v);
+
+		zvec[n] = v;
+		return true;
+		}
+
+	/**
+	 * Inserts the given value at the given index in the vector.
+	 * @param index  Which element in the vector (0-based).
+	 * @param element  The value to set the element to.
+	 */
+	void Insert(unsigned int index, ZAMValUnion& element)
+		{
+		ZVU_vec::iterator it;
+
+		if ( index < zvec.size() )
+			{
+			it = std::next(zvec.begin(), index);
+			if ( managed_yt )
+				DeleteIfManaged(index);
+			}
+		else
+			it = zvec.end();
+
+		zvec.insert(it, element);
+		}
+
+	/**
+	 * Removes the given element from the vector.
+	 * @param index  Which element in the vector (0-based).
+	 */
+	void Remove(unsigned int index)
+		{
+		if ( managed_yt )
+			DeleteIfManaged(index);
+
+		auto it = std::next(zvec.begin(), index);
+		zvec.erase(it);
+		}
+
+	/**
+	 * Resizes the vector to contain the given number of elements.
+	 * @param new_num_elements  Number of elements the vector should have.
+	 */
+	void Resize(unsigned int new_num_elements)
+		{
+		zvec.resize(new_num_elements);
+		}
+
+protected:
+	bool SetManagedElement(int n, const ZAMValUnion& v);
+	void GrowVector(int size);
+
+	void DeleteMembers();
+
+	// Deletes the given element if necessary.
+	void DeleteIfManaged(int n)
+		{
+		if ( managed_yt )
+			DeleteManagedType(zvec[n]);
+		}
+
+	// The underlying set of ZAM values.
+	ZVU_vec zvec;
+
+	// The associated main value.  A raw pointer for reasons explained
+	// above.
+	zeek::VectorVal* vv;
+
+	// The yield type of the vector elements.  Only non-nil if they
+	// are managed types.
+	zeek::TypePtr managed_yt;
+
+	// The yield type of the vector elements, whether or not it's
+	// managed.  We use a lengthier name to make sure we never
+	// confuse this with managed_yt.
+	zeek::TypePtr general_yt;
+};
+
+class ZAM_record {
+public:
+	// Similarly to ZAM_vector, we use a bare pointer for the RecordVal
+	// to simplify the memory management given the pointer cycle.
+	ZAM_record(zeek::RecordVal* _v, zeek::RecordTypePtr _rt);
+
+	~ZAM_record()
+		{
+		DeleteManagedMembers();
+		}
+
+	/**
+	 * Returns the number of fields in the record.
+	 * @return  The number of fields in the record.
+	 */
+	unsigned int Size() const	{ return zvec.size(); }
+
+	/**
+	 * Assigns the given record field to the given value.
+	 * @param field  Which field in the record to assign to.
+	 * @param v  The value to set the element to.
+	 */
+	void Assign(unsigned int field, ZAMValUnion v)
+		{
+		if ( IsInRecord(field) && IsManaged(field) )
+			Unref(zvec[field].managed_val);
+
+		zvec[field] = v;
+		is_in_record[field] = true;
+		}
+
+	/**
+	 * Provides direct raw access to one of the record's fields
```

Comment:
This style is something that would be nice to avoid, I saw in it a couple of places: Providing raw mutating access to internal class state. It breaks encapsulation and means a class can't enforce (and test) consistent properties on its internal state (memory mgmt being a case in point here).

Reply:
Okay, I think I can see how to make that work.  Check out the change I made in this regard for `ZAM_vector`, per https://github.com/zeek/zeek/commit/48585dbe324d21dd75df53c4870d74f47af5c0f4.  It allows for a few special friends to still have raw access, but no general raw access.

For `ZAM_record`'s `SetField()` (mentioned above), we can switch to having an `Assign` variant suitable for initializing empty record fields, as that's what `SetField()` is used for.  This makes code a bit less explicit, for example this current code:
```
r->SetField(1).uint_val = opcode;
```
would become:
```
r->Init(1, ZAMValUnion(opcode));
```
(where for efficiency we'd add a `ZAMValUnion constructor` that doesn't require a type, as is almost always the case when you know the type ahead of time).

How does that sound?